docs/rfcs/029-sharding-phase1.md

```diff
@@ -0,0 +1,440 @@
+that a user might create on a current-gen enterprise SSD should also work well on
+Neon. The upper bound is whatever postgres can handle: i.e. we must make sure that the
+pageserver backend is not the limiting factor in the database size_.
+
+## Non Goals
+
+- Independently distributing timelines within the same tenant. If a tenant has many
+  timelines, then sharding may be a less efficient mechanism for distributing load than
+  sharing out timelines between pageservers.
+- Distributing work in the LSN dimension: this RFC focuses on the Key dimension only,
+  based on the idea that separate mechanisms will make sense for each dimension.
+
+## Impacted Components
+
+pageserver, control plane, postgres/smgr
+
+## Terminology
+
+**Key**: a postgres page number, qualified by relation. In the sense that the pageserver is a versioned key-value store,
+the page number is the key in that store. `Key` is a literal data type in existing code.
+
+**LSN dimension**: this just means the range of LSNs (history), when talking about the range
+of keys and LSNs as a two dimensional space.
+
+## Implementation
+
+### Key sharding vs. LSN sharding
+
+When we think of sharding across the two dimensional key/lsn space, this is an
+opportunity to think about how the two dimensions differ:
+
+- Sharding the key space distributes the _write_ workload of ingesting data
+  and compacting. This work must be carefully managed so that exactly one
+  node owns a given key.
+- Sharding the LSN space distributes the _historical read_ workload. This work
+  can be done by anyone without any special coordination, as long as they can
+  see the remote index and layers.
+
+The key sharding is the harder part, and also the more urgent one, to support larger
+capacity databases. Because distributing historical LSN read work is a relatively
+simpler problem that most users don't have, we defer it to future work. It is anticipated
+that some quite simple P2P offload model will enable distributing work for historical
+reads: a node which is low on space can call out to peer to ask it to download and
+serve reads from a historical layer.
+
+### Key mapping scheme
+
+Having decided to focus on key sharding, we must next decide how we will map
+keys to shards. It is proposed to use a "wide striping" approach, to obtain a good compromise
+between data locality and avoiding entire large relations mapping to the same shard.
+
+We will define three spaces:
+
+- Key space: unsigned integer
+- Stripe space: unsigned integer
+- Shard space: integer from 0 to N-1, where we have N shards.
+
+### Stripe -> Shard mapping
+
+The main property want want from our stripe->shard mapping is
+to spread bandwidth when high throughput contiguous writes are
+happening, such as appending to a relation.
+
+Stripes are mapped to shards with a simple modulo operation:
+
+```
+Stripe | 00 | 01 | 02 | 03 | 04 | 05 | 06 | 07 | 08 | 09 | ...
+Shard  | 00 | 01 | 02 | 01 | 02 | 03 | 01 | 02 | 03 | 01 | ...
```

Comment:
I wonder if it is actually good idea to scatter stripes in this way?
So we actually use round-robin for shards. It may cause unintended and self-adjusted sync of several seqscans.
Assume that there are several huge relations in the database and that we execute OLAP queries which traverse this relations.

According to the proposed hash functions,. shard mapping will look some thing likes this:
Table A:
```
Stripe | 00 | 01 | 02 | 03 | 04 | 05 | 06 | 07 | 08 | 09 | ...
Shard  | 00 | 01 | 02 | 00 | 01 | 02 | 00 | 01 | 02 | 00 |....
```
Table B:
```
Stripe | 00 | 01 | 02 | 03 | 04 | 05 | 06 | 07 | 08 | 09 | ...
Shard  | 02 | 00 | 01 | 02 | 00 | 01 | 02 | 00 | 01 | 02 | ....
```
Table C:
```
Stripe | 00 | 01 | 02 | 03 | 04 | 05 | 06 | 07 | 08 | 09 | ...
Shard  | 01 | 02 | 00 | 01 | 02 | 00 | 01 | 02 | 00 | 01 | ....
```

So during seqscan we always access shard in this order 0 -> 1 -> 2.
Yes, there are different origins. But assume that due to some reasons some shard is overloaded with job and process requests slowly than others. It leads to "autosync" of concurrently executed seqscan queries. This slow shard will serve as barrier at which all seqscans are synchronised.

It is not always bad. For example such "self-adjustment" is used in "synchronised sequence scans" in Postgres which allows multiple seqscan queries for the same relation access reuse pages in shared buffers.

But in our case it is bad. It means that all computes will send requests to PS 0, then to PS 1, ...
So one server will be overloaded with work, while other - idle.

Reply:
Yes, this is a realistic scenario and it's worth thinking about.  Fine-tuning this kind of thing is one reason that I've built in a "layout version" -- it should be pretty easy to substitute algorithm changes in future like using the RelTag hash to select a different shard ordering for each tenant.  I'm undecided about whether we need to do that for version 1.  Our primary test cases will be the extremes of having many small relations, or scanning one huge one -- optimizing + measuring for the case of several large relations is something that may have to come later, although I agree that it's realistic.

Mitigating thoughts on the underlying behavior:
- To some extent, the "autosync" scenario is something that can happen even if we shuffle the order of nodes rather than doing identical round-robin for all relations: the load on the slow node will always be exaggerated (because all the scans will tend to linger on that node), even if we do a better job of distributing work across the other nodes by using different shard ordering for them.
- While slow nodes tend to synchronize scans, there's also a de-synchronizing force: the first node to start querying the _next_ shard will tend to race ahead.

Reply:
Sorry, I do not completely understand your arguments.
You make some special efforts to provide round-robin distribution of stripes (by excluding block number of murmur hash calculation). I do not see any advantages of such distribution: it is not expected to be more uniform or "fair" than random distribution. And the drawback of such distribution is obvious - I have explained "auto-sync" scenario.

Please notice that slow node is not something permanent. But it is system with positive feed back: once some node becomes slower than other due to any reasons, probability that more and more seqscan queries will "stuck" on this node increased. Which in turn increase workload on this node and makes it even more slower. Certainly such behaviour can happen with any distribution rule: slower node tend to become even more slower. But the question is what happen next? With round-robin next shard (shard with next id) is the best candidate for  next slowest node. If we use any other distribution, then most likely it will not happen.

I agree  that tuning distribution algorithm is endless topic. And we should start with something simple. But this is actually the main reason of my concern - you add extra complexity to hash calculation and as a result - get possible problem. While simpler solution (random distribution: including block number in murmur hash) doesn't suffer from this problem. So the question is why do we need to make extra efforts to increase probability of the problem?